### PR TITLE
[ZEPPELIN-2216] general solution to precode. refactoring jdbc precode

### PR DESCRIPTION
**Pull request by tinkoff-dwh**
### What is this PR for?
General solution to execute precode. Refactoring jdbc precode using general solution. Task contains to subtasks: executeAfterOpen, executeBeforeClose. executeBeforeClose not done because we need the context so there is a solution only for executeAfterOpen. 

### What type of PR is it?
Feature | Refactoring

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-2216

### How should this be tested?
1. Add parameter zeppelin.PySparkInterpreter.precode `someVar='text'`
2. Execute 
```
%pyspark
print(someVar)
```

### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? no

### CODE REVIEW

**Comment by tinkoff-dwh** [comment]
ready to review

**Comment by tinkoff-dwh** [comment]
prevous discussion #2096

**Comment by tinkoff-dwh** [comment]
@Leemoonsoo 
prevous discussion #2096

**Comment by Leemoonsoo** [comment]
Let me take a look in this weekend!

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterServer.java [hunk 1]
```diff
@@ -481,19 +481,24 @@ protected Object jobRun() throws Throwable {
...
 
+        InterpreterResult result = null;
+
         // Open the interpreter instance prior to calling interpret().
         // This is necessary because the earliest we can register a hook
         // is from within the open() method.
         LazyOpenInterpreter lazy = (LazyOpenInterpreter) interpreter;
         if (!lazy.isOpen()) {
           lazy.open();
+          result = lazy.executePrecode(context);
```
Comment (Leemoonsoo) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterServer.java]: Shouldn't `RemoteInterpreterServer.open()` also call `executePrecode()`?

Comment (tinkoff-dwh) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterServer.java]: it must be called once per instance of the interprete

Comment (tinkoff-dwh) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterServer.java]: @Leemoonsoo 
works same as the previous jdbc precode.
case:
   user: some_user
   JDBC interpreter (per user)
   1. there is precode `CREATE table IF NOT EXISTS tbl_#{user}`
   2. user executes `insert into tbm_some_user ...`
      - execute `precode`
      - execute paragraph text
  2. user executes `select *from tbm_some_user`
     - execute paragraph text

Spark   
1. there is precode `var someConst = 12 ...`
2. User executes `var result = 1 + someConst`
   - execute `precode`
   - execute paragraph text
3. User executes var 
```
result2 = 2 + someConts
print(result2)
```
- execute only paragraph text

Comment (Leemoonsoo) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterServer.java]: Thanks for explain.

Currently, unittest https://github.com/apache/zeppelin/pull/2221/files#diff-9c5e6fbaa5b23bc0192b4fedc12b2680R445 executes `jdbcInterpreter.executePrecode(interpreterContext);` manually after `open()`. 

Shell we have other unittest that does not call `executePrecode()` manually? so `lazy.executePrecode` in InterpreterServer can be verified.

`zeppelin-interpreter/src/test/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterTest.java` is one place the unittest actually creates RemoteInterpreter process and test against it. This place might be the place you can locate test for this feature.

Comment (tinkoff-dwh) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterServer.java]: @Leemoonsoo 
fix it, if I understand correctly your comment

**Comment by Leemoonsoo** [comment]
LGTM and merge to master if no further discussions.

**Pull request closed without merging**